Issues with undefined values of SN_MODELS

mkruckow opened the pull request:
- [x] replace undefined variable `key` by `MODEL_NAME`
- [x] check `fv` for CO_type

celiotine reviewed (changes requested):
I tried running a single binary producing this error with this branch, and a new problem has arisen because the key `fv[f'S{i+1}_{MODEL_NAME}_CO_type']`  does not exist:

```
star_1 = SingleStar(**{'mass': 103.64219775809786, 'state': 'H-rich_Core_H_burning',
                'natal_kick_array': [None, None, None, None]})
star_2 = SingleStar(**{'mass': 98.8564880638731, 'state': 'H-rich_Core_H_burning',
                'natal_kick_array': [None, None, None, None]})
binary_IF = BinaryStar(star_1, star_2, **{'time': 0.0,  'state': 'detached',  'event': 'ZAMS', 
                        'orbital_period':  1.8243881495536207, 'eccentricity': 0.0}, properties = sim_prop_IF)

binary_IF.evolve()
```

<img width="802" alt="Screen Shot 2025-01-17 at 11 24 47 AM" src="https://github.com/user-attachments/assets/c3527c17-2af4-41d3-94d4-5e0d009cb3bb" />

mkruckow commented:
Thanks for testing, @celiotine . I have made a change to instead check for the keys individually and set them with a warning to None in case they don't exist.

celiotine reviewed (changes requested):
It looks like this change is fixing the issue in step_mesa, but now there is another issue occurring in step_SN:
```
star_1 = SingleStar(**{'mass': 103.64219775809786, 'state': 'H-rich_Core_H_burning',
                'natal_kick_array': [None, None, None, None]})
star_2 = SingleStar(**{'mass': 98.8564880638731, 'state': 'H-rich_Core_H_burning',
                'natal_kick_array': [None, None, None, None]})
binary_IF = BinaryStar(star_1, star_2, **{'time': 0.0,  'state': 'detached',  'event': 'ZAMS', 
                        'orbital_period':  1.8243881495536207, 'eccentricity': 0.0}, properties = sim_prop_IF)

binary_IF.evolve()
```
<img width="814" alt="Screen Shot 2025-01-20 at 1 01 20 PM" src="https://github.com/user-attachments/assets/daadb930-c7af-44b9-8eac-ab01eb795ff0" />

mkruckow commented:
Sure, the new issue happened because of the `None` type not being eligible for `np.isnan`, hence I changed them all to `pd.isna` as we already said some time ago, but never penetrated to all cases in step_SN.
@celiotine can you have another check? Thanks for all the testing.

celiotine reviewed (dismissed):
Ok, now it looks like there is another type issue in the CO_radius function... hopefully this is the last one!

```
star_1 = SingleStar(**{'mass': 103.64219775809786, 'state': 'H-rich_Core_H_burning',
                'natal_kick_array': [None, None, None, None]})
star_2 = SingleStar(**{'mass': 98.8564880638731, 'state': 'H-rich_Core_H_burning',
                'natal_kick_array': [None, None, None, None]})
binary_IF = BinaryStar(star_1, star_2, **{'time': 0.0,  'state': 'detached',  'event': 'ZAMS', 
                        'orbital_period':  1.8243881495536207, 'eccentricity': 0.0}, properties = sim_prop_IF)

binary_IF.evolve()
```

<img width="809" alt="Screen Shot 2025-01-21 at 10 41 40 AM" src="https://github.com/user-attachments/assets/46c00c7e-e4cb-47ec-97d9-d4b5c6943be4" />

mkruckow commented:
Here we run into the problem, that the function expects a float type while a None type in input. See issue #236 .
I now replaced the undefined values instead with `np.nan`, this will work for all functions expecting floats.
A warning: this might raise issues in the future, when other non float values get added to the list, hence I added a todo to check for this.

celiotine reviewed (dismissed):
Now it looks like a similar issue has propagated to the detached step:
<img width="805" alt="Screen Shot 2025-01-22 at 2 04 49 PM" src="https://github.com/user-attachments/assets/6f83907c-251b-4c06-a716-44d007671b35" />

mkruckow commented:
This more looks like a problem in step SN not causing the binary to fail but giving values to step detached, which can't be handled there.
So, I now set the Model to None, like no values are available.

celiotine reviewed (dismissed):
This is still not working for me:
<img width="812" alt="Screen Shot 2025-02-03 at 1 23 47 PM" src="https://github.com/user-attachments/assets/76aabdc5-d96e-4139-b73a-0f2c2aa52754" />
```
star_1 = SingleStar(**{'mass': 118.22761412999836, 'state': 'H-rich_Core_H_burning',
                'natal_kick_array': [None, None, None, None]})
star_2 = SingleStar(**{'mass': 109.48134586589462, 'state': 'H-rich_Core_H_burning',
                'natal_kick_array': [None, None, None, None]})
binary_IF = BinaryStar(star_1, star_2, **{'time': 0.0,  'state': 'detached',  'event': 'ZAMS', 
                        'orbital_period':  1.7068750446542835, 'eccentricity': 0.0}, properties = sim_prop_IF)

binary_IF.evolve()
```

mkruckow commented:
Sorry, I missed, that the previously set values overwrite the `None` after the `break`.

celiotine reviewed (dismissed):
Ok, it looks like the issue is now fixed for some binaries that were previously failing with the original error, however, the PR is now introducing a new error:

```
Traceback (most recent call last):
  File "/home/cel2717/.conda/envs/posydon-source/lib/python3.11/site-packages/posydon/popsyn/binarypopulation.py", line 327, in _safe_evolve
    binary.evolve()
  File "/home/cel2717/.conda/envs/posydon-source/lib/python3.11/site-packages/posydon/binary_evol/binarystar.py", line 222, in evolve
    self.run_step()
  File "/home/cel2717/.conda/envs/posydon-source/lib/python3.11/site-packages/posydon/binary_evol/binarystar.py", line 249, in run_step
    next_step(self)
  File "/home/cel2717/.conda/envs/posydon-source/lib/python3.11/site-packages/posydon/binary_evol/DT/step_detached.py", line 1169, in __call__
    interp1d_pri = get_star_data(
                   ^^^^^^^^^^^^^^
  File "/home/cel2717/.conda/envs/posydon-source/lib/python3.11/site-packages/posydon/binary_evol/DT/step_detached.py", line 1153, in get_star_data
    interp1d["mass"] = PchipInterpolator(age, kvalue["mass"])
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/cel2717/.conda/envs/posydon-source/lib/python3.11/site-packages/scipy/interpolate/_cubic.py", line 234, in __init__
    x, _, y, axis, _ = prepare_input(x, y, axis)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/cel2717/.conda/envs/posydon-source/lib/python3.11/site-packages/scipy/interpolate/_cubic.py", line 54, in prepare_input
    raise ValueError("`y` must contain only finite values.")
ValueError: `y` must contain only finite values.

Failed Binary Initial Conditions:
S1 mass: 7.12421327515828
S2 mass: 3.259588216276884
S1 state: H-rich_Core_H_burning
S2 state: H-rich_Core_H_burning
orbital period: 15.60832398246785
eccentricity: 0.0
binary state: detached
binary event: detached
S1 natal kick array: [None None None None]
S2 natal kick array: [None None None None]
```

mkruckow commented:
This was an issue, because `np.isnan` and `pd.isna` return different boolean types, which behave differently with the tilde.

celiotine reviewed (approved):
looks good to me!
tested on previously failed binaries and ran a 10k population with no issues
